File: 13.DFS.BFS/783._Minimum_Risk_Path.py
#  https://www.lintcode.com/problem/minimum-risk-path/

import heapq
# 构建
# heapq.heapify(heap)
# pop
# cur = heapq.heappop(heap)
# push
# heapq.heappush(heap, cur.next)

class Solution:
    """
    @param n: maximum index of position.
    @param m: the number of undirected edges.
    @param x: 
    @param y: 
    @param w: 
    @return: return the minimum risk value.
    """
    def getMinRiskValue(self, n, m, x, y, w):
        G = []
        for i in range(n+1):
          G.append([])
        
        # 构建图
        for i in range(m):
          a = x[i]
          b = y[i]
          c = w[i]
          G[a].append([b,c])
          G[b].append([a,c])
        
        Q = []

        INF = 2147483647
        risk = [INF] * (n+1)
        risk[0] = 0

        # push <risk, node>
        heapq.heappush(Q, [0, 0])

        while len(Q) > 0:
          cRisk, cNode = heapq.heappop(Q)
          # 拿当前结点的后续
          for nextInfo in G[cNode]:
            next = nextInfo[0]
            nextRisk = max(cRisk, nextInfo[1])
            if nextRisk < risk[next]:
              risk[next] = nextRisk
              heapq.heappush(Q, [nextRisk, next])
        
        return risk[n]

File: 13.DFS.BFS/783._Minimum_Risk_Path.cpp
// https://www.lintcode.com/problem/minimum-risk-path/

#include <assert.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <algorithm>
#include <iostream>
#include <numeric>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 public:
  /**
   * @param n: maximum index of position.
   * @param m: the number of undirected edges.
   * @param x: undirected edges
   * @param y: undirected edges
   * @param w: undirected edges risk
   * @return: return the minimum risk value.
   */
  int getMinRiskValue(int n, int m, vector<int> &x, vector<int> &y,
                      vector<int> &w) {
    vector<vector<pair<int, int>>> G(n + 1);

    for (int i = 0; i < m; i++) {
      const int a = x[i], b = y[i], c = w[i];
      G[a].push_back({b, c});
      G[b].push_back({a, c});
    }

    vector<int> risk(n + 1, INT_MAX);
    // Q 里面记录了走到某个点的最小代价
    // 每次出队的时候，我们总是出队最小的
    auto cmp = [&](int a, int b) { return risk[a] > risk[b]; };
    priority_queue<int, vector<int>, decltype(cmp)> Q(cmp);
    // 将起始点放到队列中
    Q.push(0);
    // 一开始出发点risk为0
    risk[0] = 0;

    while (!Q.empty()) {
      auto cur = Q.top();
      Q.pop();

      // 查看当前结点的后继
      for (auto nextValue : G[cur]) {
        const int curRisk = risk[cur];
        const int nextRisk = max(curRisk, nextValue.second);
        if (nextRisk < risk[nextValue.first]) {
          risk[nextValue.first] = nextRisk;
          Q.push(nextValue.first);
        }
      }
    }

    return risk[n];
  }
};